app/uk/gov/hmrc/economiccrimelevyregistration/controllers/RegistrationValidationController.scala

```diff
@@ -16,50 +16,61 @@
 import uk.gov.hmrc.play.bootstrap.backend.controller.BackendController
 
 import javax.inject.{Inject, Singleton}
-import scala.concurrent.{ExecutionContext, Future}
+import scala.concurrent.ExecutionContext
 
 @Singleton()
 class RegistrationValidationController @Inject() (
   cc: ControllerComponents,
-  registrationRepository: RegistrationRepository,
+  registrationService: RegistrationService,
   authorise: AuthorisedAction,
   registrationValidationService: RegistrationValidationService,
   registrationAdditionalInfoService: RegistrationAdditionalInfoService
 )(implicit ec: ExecutionContext)
     extends BackendController(cc)
-    with Logging {
+    with Logging
+    with BaseController
+    with ErrorHandler {
 
-  def getValidationErrors(id: String): Action[AnyContent] = authorise.async { _ =>
-    registrationRepository.get(id).flatMap {
-      case Some(registration) =>
-        registrationAdditionalInfoService
-          .get(registration.internalId)
-          .fold(
-            error => {
-              logger.error(
-                s"Failed to find additional information for amendment with internal id: ${registration.internalId}"
-              )
-              NotFound
-            },
-            additionalInfo =>
-              registrationValidationService.validateRegistration(registration, additionalInfo) match {
-                case Valid(_)   => NoContent
-                case Invalid(e) => Ok(Json.toJson(DataValidationErrors(e.toList)))
-              }
-          )
-      case None               => Future.successful(NotFound)
-    }
+  def checkForValidationErrors(id: String): Action[AnyContent] = authorise.async { _ =>
+    (for {
+      registration   <- registrationService.getRegistration(id).asResponseError
+      additionalInfo <- registrationAdditionalInfoService.get(registration.internalId).asResponseError
+      result          = resolveRegistrationValidationExecutionPath(registration, additionalInfo)
+    } yield result).convertToResult(OK)
   }
 
+  private def resolveRegistrationValidationExecutionPath(
+    registration: Registration,
+    additionalInfo: RegistrationAdditionalInfo
+  ): Boolean                                                                                                           =
+    if (registration.isRegistration) {
+      checkForErrorsInRegistrationValidation(registrationValidationService.validateRegistration(registration))
+    } else {
+      checkForErrorsInSubscriptionValidation(
+        registrationValidationService.validateSubscription(registration, additionalInfo)
+      )
+    }
+  private def checkForErrorsInRegistrationValidation(registration: Either[DataValidationError, Registration]): Boolean =
+    registration match {
+      case Left(_)  => true
+      case Right(_) => false
+    }
+
+  private def checkForErrorsInSubscriptionValidation(
```

Comment:
Here we will be returning a boolean in our result? Why? If there are no errors we can just return a unit here, not write anything in the body and return OK from the controller

Reply:
Yup, that makes sense, I've had to do that because at the moment didn't know how to bring down validateRegistration & validateSubscribtion method results, which return two different types, to the same type. Based on your previous comment regarding changing Either to EitherT in validationService now I'm able to remove polarization of those two methods more elegantly. Thanks for the tip 😄